### ISSUE TITLE
Insufficient feedback when submitting offers that are discarded due to rounding errors

### ISSUE THREAD

**Issue by vcarl** [stellar/stellar-protocol#195]
When submitting tiny offers of `0.0000001`, an offer does not end up on the books. jonjove suggested in Keybase chat that it may be due to CAP-0004 removing them due to rounding errors, which makes sense. However there doesn't appear to be a way to detect when this occurs.

Our users are ending up with this tiny balance likely through rounding errors when selling their entire balance. When they try to sell this "dust," the UI tells them that the trade was successfully submitted and immediately filled, because this vanishing offer behavior is identical to that case. 

As an application developer, it would be helpful to have some kind of feedback when this happens, or at least a way to check if it has. Right now all information I can access from Horizon looks the same as if the offer was immediately filled, so the best I can do is prevent users from being able to submit trades that I've confirmed will always fail.

I have an [envelope XDR](https://www.stellar.org/laboratory/#xdr-viewer?input=AAAAAMdXtr4aSKxUz0fCJMfqM2zoTr3powMu9Bo%2Be0z81VgVAAAAZAEaLvMAAAIGAAAAAAAAAAAAAAABAAAAAAAAAAMAAAAAAAAAAVdTRAAAAAAA5VsQH718zt2rcOs3iMwCZtCvIbuK2A9ha9uALJgLZ%2BEAAAAAAAAABAAHoSAAIGy3AAAAAAAAAAAAAAAAAAAAAfzVWBUAAABAxBKVqq01%2BEmOBZosWpiLyviBrfTGyBi7Q%2BKc7h5utTMP0l6ukY6lXej7gC5LVz3t3Bbl8PEwqGASKPUhrBRrBA%3D%3D%0A&type=TransactionEnvelope&network=public) and [result XDR](https://www.stellar.org/laboratory/#xdr-viewer?input=AAAAAAAAAGQAAAAAAAAAAQAAAAAAAAADAAAAAAAAAAAAAAACAAAAAA%3D%3D&type=TransactionResult&network=public) handy for one of these transaction submissions.

**Comment by MonsieurNicolas** [comment]
Hi

You need to look in the transaction result, the `manageOfferResult` field gives information on what happened.

In your example:
```
offersClaimed: Array[0]
offer: [manageOfferDeleted]
```
means that the offer was deleted (meaning never created in this case), if trades occur as part of the operation you would see the amounts here.

Even in cases where the offer gets created/modified, you want to check the content of `manageOfferResult` that will tell you the actual amount for the offer in the order book at the end of the operation (in case of adjustments because of trades, liabilities or rounding).

**Comment by vcarl** [comment]
Ah, that's a good point. I do have access to the XDR from our app code, I should be able to check for that case.

OTOH an offer being immediately deleted upon submission doesn't seem like a "success" case. It would be nice if this appeared as an error instead, but that might be better solved at the SDK level.

**Comment by vcarl** [comment]
[This offer result](https://www.stellar.org/laboratory/#xdr-viewer?input=AAAAAAAAAGQAAAAAAAAAAQAAAAAAAAADAAAAAAAAAAEAAAAAp3h16puj5nLJcgEdX0SQQfdTu5pfKAwqsZiWTNPLDYYAAAAAAlm23QAAAAFVU0QAAAAAAOimGoYeYK9g%2BAdz4GNG5ccsvlncrdo3YI1Y70JRHZ%2FcAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAACAAAAAA%3D%3D&type=TransactionResult&network=public) never created an offer, yet it claims that there's 1 offer claimed. [My account's transaction history](https://horizon.stellar.org/accounts/GDDVPNV6DJEKYVGPI7BCJR7KGNWOQTV55GRQGLXUDI7HWTH42VMBK6MA/transactions) doesn't have a matching record. Likewise, [the transaction meta](https://www.stellar.org/laboratory/#xdr-viewer?input=AAAAAQAAAAIAAAADATm7WwAAAAAAAAAAx1e2vhpIrFTPR8Ikx%2BozbOhOvemjAy70Gj57TPzVWBUAAAAAMGYnMQEaLvMAAAI1AAAADgAAAAEAAAAAhD8BLsZFQEF33rKS6YopQUT3b6iLBG4nspe68%2FDBNBYAAAAAAAAAAAEAAAAAAAAAAAAAAQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEBObtbAAAAAAAAAADHV7a%2BGkisVM9HwiTH6jNs6E696aMDLvQaPntM%2FNVYFQAAAAAwZicxARou8wAAAjYAAAAOAAAAAQAAAACEPwEuxkVAQXfespLpiilBRPdvqIsEbieyl7rz8ME0FgAAAAAAAAAAAQAAAAAAAAAAAAABAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAQAAAAYAAAADATm7PAAAAAAAAAAAp3h16puj5nLJcgEdX0SQQfdTu5pfKAwqsZiWTNPLDYYAAAAB1ftlJwD3w1UAAATQAAAALAAAAAEAAAAAxHHGQ3BiyVBqiTQuU4oa2kBNL0HPHTolX0Mh98bg4XUAAAAAAAAACWxvYnN0ci5jbwAAAAEAAAAAAAAAAAAAAQAAAADTarmrAAAAAZjZfQUAAAAAAAAAAAAAAAEBObtbAAAAAAAAAACneHXqm6PmcslyAR1fRJBB91O7ml8oDCqxmJZM08sNhgAAAAHV%2B2UnAPfDVQAABNAAAAAsAAAAAQAAAADEccZDcGLJUGqJNC5TihraQE0vQc8dOiVfQyH3xuDhdQAAAAAAAAAJbG9ic3RyLmNvAAAAAQAAAAAAAAAAAAABAAAAANNquasAAAABmNl9BQAAAAAAAAAAAAAAAwE5uzwAAAABAAAAAKd4deqbo%2BZyyXIBHV9EkEH3U7uaXygMKrGYlkzTyw2GAAAAAVVTRAAAAAAA6KYahh5gr2D4B3PgY0blxyy%2BWdyt2jdgjVjvQlEdn9wAAAAAAmJaAH%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FAAAAAQAAAAEAAAAAAAAAAAAAAAACYloAAAAAAAAAAAAAAAABATm7WwAAAAEAAAAAp3h16puj5nLJcgEdX0SQQfdTu5pfKAwqsZiWTNPLDYYAAAABVVNEAAAAAADophqGHmCvYPgHc%2BBjRuXHLL5Z3K3aN2CNWO9CUR2f3AAAAAACYloAf%2F%2F%2F%2F%2F%2F%2F%2F%2F8AAAABAAAAAQAAAAAAAAAAAAAAAAJiWgAAAAAAAAAAAAAAAAMBObs8AAAAAgAAAACneHXqm6PmcslyAR1fRJBB91O7ml8oDCqxmJZM08sNhgAAAAACWbbdAAAAAVVTRAAAAAAA6KYahh5gr2D4B3PgY0blxyy%2BWdyt2jdgjVjvQlEdn9wAAAAAAAAAAAJiWgAAAAAJAAAAAgAAAAAAAAAAAAAAAAAAAAEBObtbAAAAAgAAAACneHXqm6PmcslyAR1fRJBB91O7ml8oDCqxmJZM08sNhgAAAAACWbbdAAAAAVVTRAAAAAAA6KYahh5gr2D4B3PgY0blxyy%2BWdyt2jdgjVjvQlEdn9wAAAAAAAAAAAJiWgAAAAAJAAAAAgAAAAAAAAAAAAAAAA%3D%3D&type=TransactionMeta&network=public) also shows an amount in "ledgerEntryUpdated".

The js-stellar-sdk doesn't expose the contents of these in any way that's documented, I've had to find my way around the source of several different repos in order to figure out how to do decode and inspect the contents. The more I poke at this, the more it looks like a bug somewhere. This example in particular appears to include information that it shouldn't, which suggests a bug in core.

At the moment, I'm not sure it's possible for me to detect when these offers are deleted without incorrectly triggering on successful trades that are filled immediately.

**Comment by MonsieurNicolas** [comment]
The sold/bought amount in the meta is "0", so basically nothing happened with that trade.

Horizon does a bunch of filtering which in this case makes it hard to understand. Do you have the XDR for the transaction that resulted in this result?

Also it would be very useful to have the XDR for the corresponding trustlines right before you submit the transaction (a dump of the output of the accounts endpoint like https://horizon.stellar.org/accounts/GDDVPNV6DJEKYVGPI7BCJR7KGNWOQTV55GRQGLXUDI7HWTH42VMBK6MA )

**Comment by vcarl** [comment]
Sorry, I was on vacation last week. Here's the [account balances](https://gist.github.com/vcarl/045cd9b7b2c73d576e5197412a2ecbd8) before the trade was submitted.

[Here's the envelope as submitted]( https://www.stellar.org/laboratory/#xdr-viewer?input=AAAAAA2pqoHPAI9DNk4YCcmzQLzf8WIctcpwsZkZVZmKUyaqAAAAZAEeqiMAAAA4AAAAAAAAAAAAAAABAAAAAAAAAAMAAAABV1NEAAAAAADlWxAfvXzO3atw6zeIzAJm0K8hu4rYD2Fr24AsmAtn4QAAAAAAAAAAAAAAAQKCZUkAmJaAAAAAAAAAAAAAAAAAAAAAAA%3D%3D&type=TransactionEnvelope&network=public), to submit a sell order of `.0000001` WSD, which I had an existing trustline and balance for.

And here's the response from the network:
[envelope](https://www.stellar.org/laboratory/#xdr-viewer?input=AAAAAA2pqoHPAI9DNk4YCcmzQLzf8WIctcpwsZkZVZmKUyaqAAAAZAEeqiMAAAA4AAAAAAAAAAAAAAABAAAAAAAAAAMAAAABV1NEAAAAAADlWxAfvXzO3atw6zeIzAJm0K8hu4rYD2Fr24AsmAtn4QAAAAAAAAAAAAAAAQKCZUkAmJaAAAAAAAAAAAAAAAAAAAAAAYpTJqoAAABATv6yezN7ri%2B4L%2FINc7jtFBVLxV4MbY6idhtGVEZL1WOuKPNziT04lv0WO90%2BbUl8E4nFp0N3iWjn4O%2FEs0ZwAQ%3D%3D&type=TransactionEnvelope&network=public)
[result](https://www.stellar.org/laboratory/#xdr-viewer?input=AAAAAAAAAGQAAAAAAAAAAQAAAAAAAAADAAAAAAAAAAEAAAAAHvYQGA2d%2Fiq3Jer20b8b9%2FfC%2Fj5KZ%2FZ8646b5oxTAesAAAAAAn%2FUPwAAAAAAAAAAAAAAAAAAAAFXU0QAAAAAAOVbEB%2B9fM7dq3DrN4jMAmbQryG7itgPYWvbgCyYC2fhAAAAAAAAAAAAAAACAAAAAA%3D%3D&type=TransactionResult&network=public)
[result meta](https://www.stellar.org/laboratory/#xdr-viewer?input=AAAAAQAAAAIAAAADATx0%2FgAAAAAAAAAADamqgc8Aj0M2ThgJybNAvN%2FxYhy1ynCxmRlVmYpTJqoAAAAAAiVQRQEeqiMAAAA3AAAAAgAAAAAAAAAAAAAAAAEAAAAAAAAAAAAAAQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEBPHT%2BAAAAAAAAAAANqaqBzwCPQzZOGAnJs0C83%2FFiHLXKcLGZGVWZilMmqgAAAAACJVBFAR6qIwAAADgAAAACAAAAAAAAAAAAAAAAAQAAAAAAAAAAAAABAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAQAAAAYAAAADATxoowAAAAAAAAAAHvYQGA2d%2Fiq3Jer20b8b9%2FfC%2Fj5KZ%2FZ8646b5oxTAesAAAAiB2Qi6wEtRR8AAAEDAAAACQAAAAEAAAAAhD8BLsZFQEF33rKS6YopQUT3b6iLBG4nspe68%2FDBNBYAAAAAAAAAASAAAAABAAAAAAAAAAAAAAEAAAAABwg7pgAAAAH6TH%2FRAAAAAAAAAAAAAAABATx0%2FgAAAAAAAAAAHvYQGA2d%2Fiq3Jer20b8b9%2FfC%2Fj5KZ%2FZ8646b5oxTAesAAAAiB2Qi6wEtRR8AAAEDAAAACQAAAAEAAAAAhD8BLsZFQEF33rKS6YopQUT3b6iLBG4nspe68%2FDBNBYAAAAAAAAAASAAAAABAAAAAAAAAAAAAAEAAAAABwg7pgAAAAH6TH%2FRAAAAAAAAAAAAAAADATxoowAAAAEAAAAAHvYQGA2d%2Fiq3Jer20b8b9%2FfC%2Fj5KZ%2FZ8646b5oxTAesAAAABV1NEAAAAAADlWxAfvXzO3atw6zeIzAJm0K8hu4rYD2Fr24AsmAtn4QAAAAAAAAAAf%2F%2F%2F%2F%2F%2F%2F%2F%2F8AAAABAAAAAQAAAAA9ZI2AAAAAAAAAAAAAAAAAAAAAAAAAAAEBPHT%2BAAAAAQAAAAAe9hAYDZ3%2BKrcl6vbRvxv398L%2BPkpn9nzrjpvmjFMB6wAAAAFXU0QAAAAAAOVbEB%2B9fM7dq3DrN4jMAmbQryG7itgPYWvbgCyYC2fhAAAAAAAAAAB%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwAAAAEAAAABAAAAAD1kjYAAAAAAAAAAAAAAAAAAAAAAAAAAAwE8aKMAAAACAAAAAB72EBgNnf4qtyXq9tG%2FG%2Ff3wv4%2BSmf2fOuOm%2BaMUwHrAAAAAAJ%2F1D8AAAAAAAAAAVdTRAAAAAAA5VsQH718zt2rcOs3iMwCZtCvIbuK2A9ha9uALJgLZ%2BEAAAABAnbAXwCYloACgmVJAAAAAAAAAAAAAAAAAAAAAQE8dP4AAAACAAAAAB72EBgNnf4qtyXq9tG%2FG%2Ff3wv4%2BSmf2fOuOm%2BaMUwHrAAAAAAJ%2F1D8AAAAAAAAAAVdTRAAAAAAA5VsQH718zt2rcOs3iMwCZtCvIbuK2A9ha9uALJgLZ%2BEAAAABAnbAXwCYloACgmVJAAAAAAAAAAAAAAAA&type=TransactionMeta&network=public)

Here's another for a trade that doesn't fill immediately

[submitted envelope](https://www.stellar.org/laboratory/#xdr-viewer?input=AAAAAA2pqoHPAI9DNk4YCcmzQLzf8WIctcpwsZkZVZmKUyaqAAAAZAEeqiMAAAA5AAAAAAAAAAAAAAABAAAAAAAAAAMAAAAAAAAAAVdTRAAAAAAA5VsQH718zt2rcOs3iMwCZtCvIbuK2A9ha9uALJgLZ%2BEAAAAAAAAABAAAAAEAAAAEAAAAAAAAAAAAAAAAAAAAAA%3D%3D&type=TransactionEnvelope&network=public)

response:
[envelope](https://www.stellar.org/laboratory/#xdr-viewer?input=AAAAAA2pqoHPAI9DNk4YCcmzQLzf8WIctcpwsZkZVZmKUyaqAAAAZAEeqiMAAAA5AAAAAAAAAAAAAAABAAAAAAAAAAMAAAAAAAAAAVdTRAAAAAAA5VsQH718zt2rcOs3iMwCZtCvIbuK2A9ha9uALJgLZ%2BEAAAAAAAAABAAAAAEAAAAEAAAAAAAAAAAAAAAAAAAAAYpTJqoAAABAnZ%2BqSVGrTw0OX5mJ6PD%2Ftmb4v6Qv2mmQTJ9B6p2fqZrl4sMACCXaMveTX57POG66sehmkUXIq9z99NmTJZF9DQ%3D%3D&type=TransactionEnvelope&network=public)
[result](https://www.stellar.org/laboratory/#xdr-viewer?input=AAAAAAAAAGQAAAAAAAAAAQAAAAAAAAADAAAAAAAAAAAAAAAAAAAAAA2pqoHPAI9DNk4YCcmzQLzf8WIctcpwsZkZVZmKUyaqAAAAAAKAc0kAAAAAAAAAAVdTRAAAAAAA5VsQH718zt2rcOs3iMwCZtCvIbuK2A9ha9uALJgLZ%2BEAAAAAAAAABAAAAAEAAAAEAAAAAAAAAAAAAAAA&type=TransactionResult&network=public)
[result meta](https://www.stellar.org/laboratory/#xdr-viewer?input=AAAAAQAAAAIAAAADATx1eAAAAAAAAAAADamqgc8Aj0M2ThgJybNAvN%2FxYhy1ynCxmRlVmYpTJqoAAAAAAiVP4QEeqiMAAAA4AAAAAgAAAAAAAAAAAAAAAAEAAAAAAAAAAAAAAQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEBPHV4AAAAAAAAAAANqaqBzwCPQzZOGAnJs0C83%2FFiHLXKcLGZGVWZilMmqgAAAAACJU%2FhAR6qIwAAADkAAAACAAAAAAAAAAAAAAAAAQAAAAAAAAAAAAABAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAQAAAAUAAAAAATx1eAAAAAIAAAAADamqgc8Aj0M2ThgJybNAvN%2FxYhy1ynCxmRlVmYpTJqoAAAAAAoBzSQAAAAAAAAABV1NEAAAAAADlWxAfvXzO3atw6zeIzAJm0K8hu4rYD2Fr24AsmAtn4QAAAAAAAAAEAAAAAQAAAAQAAAAAAAAAAAAAAAAAAAADATx1eAAAAAAAAAAADamqgc8Aj0M2ThgJybNAvN%2FxYhy1ynCxmRlVmYpTJqoAAAAAAiVP4QEeqiMAAAA5AAAAAgAAAAAAAAAAAAAAAAEAAAAAAAAAAAAAAQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEBPHV4AAAAAAAAAAANqaqBzwCPQzZOGAnJs0C83%2FFiHLXKcLGZGVWZilMmqgAAAAACJU%2FhAR6qIwAAADkAAAADAAAAAAAAAAAAAAAAAQAAAAAAAAAAAAABAAAAAAAAAAAAAAAAAAAABAAAAAAAAAAAAAAAAwE8b8sAAAABAAAAAA2pqoHPAI9DNk4YCcmzQLzf8WIctcpwsZkZVZmKUyaqAAAAAVdTRAAAAAAA5VsQH718zt2rcOs3iMwCZtCvIbuK2A9ha9uALJgLZ%2BEAAAAAAEFAkH%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FAAAAAQAAAAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAABATx1eAAAAAEAAAAADamqgc8Aj0M2ThgJybNAvN%2FxYhy1ynCxmRlVmYpTJqoAAAABV1NEAAAAAADlWxAfvXzO3atw6zeIzAJm0K8hu4rYD2Fr24AsmAtn4QAAAAAAQUCQf%2F%2F%2F%2F%2F%2F%2F%2F%2F8AAAABAAAAAQAAAAAAAAABAAAAAAAAAAAAAAAAAAAAAA%3D%3D&type=TransactionMeta&network=public)

I can gather specific cases if that would be helpful. So far I have not identified any way that the response returned by these trades is different from successful trades.

**Comment by jonjove** [comment]
@vcarl I agree that properly interpreting a `ManageOfferResult` is not easy, but I believe that all of the requisite information is there. I'm going to try to walk you through it in the case where the operation succeeded. These comments are intended to apply for protocol version 10. First, let's discuss the `effect`:

- `MANAGE_OFFER_CREATED`: A new offer was created and it remained in the book after the operation was applied
- `MANAGE_OFFER_UPDATED`: The offer specified in `ManageOfferOp.offerID` was updated and it remained in the book after the operation was applied
- `MANAGE_OFFER_DELETED`: Either a new offer was created or the offer specified in `ManageOfferOp.offerID` was updated, but it did not remain in the book after the operation was applied

It is important to note that the `effect` provides _no_ information about whether or not any offers were executed. For example, a `ManageOffer` operation could be constructed to create an offer. When this operation is applied, it could occur that the new offer does not cross any existing offer but also cannot be adjusted (see CAP-0004). Since the offer cannot be adjusted, it is removed from the book. Using the guidelines from above, the `effect` is then `MANAGE_OFFER_DELETED` although no offers were executed.

Since the `effect` cannot be used to determine whether any offers were executed, one must next turn to `offersClaimed`. Some properties of `offersClaimed` are:

- `offersClaimed` is not empty if and only if applying the `ManageOffer` operation led to an intermediate state with a crossed book
- Each offer contained in `offersClaimed` _except the last_ is guaranteed to be removed from the book. The last offer may have been partially executed, or not even executed at all due to the error threshold (see CAP-0004)
- Every offer contained in `offersClaimed` will have the same `assetSold`. The total amount bought by the source of the `ManageOffer` operation is the sum of `amountSold` aggregated over all offers contained in `offersClaimed`
- Every offer contained in `offersClaimed` will have the same `assetBought`. The total amount sold by the source of the `ManageOffer` operation is the sum of `amountBought` aggregated over all offers contained in `offersClaimed`

With all of the above in mind, let's turn our attention to some of the examples you shared:

1. The first example from your most recent comment is equivalent to my example above: The created offer did not cross the book (`offersClaimed` was empty) but did not remain in the book after the operation was applied (`effect == manageOfferDeleted`)
2. The second example from your most recent comment can be interpreted as follows: The created offer did not cross the book (`offersClaimed` was empty) but did remain in the book after the operation was applied (`effect == manageOfferCreated`)
3. The example of a created offer that has 1 offer claimed: The created offer did cross the book but nothing was exchanged (`amountSold == amountBought == 0`) and the offer did not remain in the book after the operation was applied (`effect == manageOfferDeleted`)

There is nothing ambiguous about any of the above results if you apply these guidelines. It is important to recognize that the sizes of created and updated offers can be different from what was submitted in the `ManageOffer` operation because the offer (partially or completely) executed, because the offer was adjusted, or both. If you come across a result that you think cannot be interpreted using the guidelines above, feel free to ask about it.

**Comment by vcarl** [comment]
This is the only part of the application where we have to read XDRs to get the information we need, so I'm going to continue this conversation further downstream. It may be possible on a technical level to determine when this occurs, but I don't believe this is a reasonable level of detail to expect an average web application to handle. If not here, then either Horizon or the SDK should determine when this occurs.

**Comment by morleyzhi** [comment]
@MonsieurNicolas In this case, why does the operation succeed? Shouldn't it fail, and thus fail the transaction?

If I'm a developer and I create an offer, I'd reasonably expect it to appear on the orderbook (or to fill). If it does neither, is that really a successful operation?

@vcarl pointed out that if the offer immediately fills, it also has `manageOfferResult` set to `MANAGE_OFFER_DELETED`. If that's true, it's not possible to tell the difference between an offer that successfully filled and an offer that was deleted immediately.

I think it's better to mark these small-amount, immediately-deleted offers as failed. Otherwise, everyone who submits offers has to inspect the XDR to know if they succeeded -- and if Carl is right, then it's not even reliable to do that.

**Comment by MonsieurNicolas** [comment]
We debated failing for the special case of an offer not getting created because of rounding issues, but what we landed on was that:
* We don't want people to implement code that is 90% correct in SDKs, we want to encourage people to deal with the "edge cases" by surfacing them right away
    * it's actually not that different from an offer getting created with a different amount specified or from other trades happening even though the offer doesn't get created: people have to check the return value by verifying both trades (`ClaimOfferAtom`) and effective amounts (trades and offer created).
* We don't want to fail operations if they are "well formed", that way people can submit batches of trades (multiple ops) in one transaction and don't have to worry about some of those trades being only partial and/or fall victim of rounding.

On a separate note, all the information is there, @jonjove wrote a pretty detailed description. You should re-review it, and pay particular attention to what the `offersClaimed` field does to help you.

**Comment by vcarl** [comment]
I believe from @jonjove's last post there is indeed a method for determining when a transaction has been removed for this reason. However the length of his post and the enduring confusion around this I think is reason enough to avoid burdening clients with this. When writing an app 2 layers of abstraction from the protocol, I don't expect to have to dive into protocol details to confirm it's working the way I want.

> We don't want people to implement code that is 90% correct in SDKs, we want to encourage people to deal with the "edge cases" by surfacing them right away

I'm not sure I agree this is surfaced right away @MonsieurNicolas. It took hours of investigation for me to uncover what was happening in StellarX and discussion with protocol devs to identify why. Despite @jonjove's assistence I ended up "solving" this by adding a special case to block trades that are below a threshold—which surely shows some false positives.

The decision to treat this as a successful transaction puts a significant burden on clients that want to provide a good experience to their users. For the vast majority of users, this situation is a failed transaction. It seems to me that treating it as successful prioritizes an edge case in a fringe use case over typical usage of the protocol.

**Comment by vcarl** [comment]
> Each offer contained in offersClaimed except the last is guaranteed to be removed from the book. The last offer may have been partially executed, or not even executed at all due to the error threshold

From @jonjove's post, I believe this ambiguity makes it impossible to confirm when an offer has been ignored due to the rounding error here. Offers that are partially claimed will have the same result as an offer than is discarded due to rounding, but as a client I need to handle them differently.

**Comment by MonsieurNicolas** [comment]
Maybe we can augmented results in Horizon if we want an easy way for SDKs to deal with this?
I thought the "trades" endpoint was supposed to help with this?

The issue with the current contract is that rounding has arbitrary effect on what you would consider a "good" or "bad" trade, it all depends on context and what is being traded.
Note that rounding is a real problem for wallets that deal with things like BTC vs USD for example.

The second question is something different:
I'll let @jonjove give you specifics on `offersClaimed`, my understanding of it, is that the amounts in there (sold/bought) are real (as in, if it says "amountSold = 1" "1" was taken from the offer, "0" means nothing happened) and I would not use that to derive order book state.
If you want to build a current view of the order book, you have to call Horizon as other transactions could have changed the order book in general (that said, I don't know if there is a simple "streaming"/"differential" order book type of API to fetch specific trading pairs).

**Comment by morleyzhi** [comment]
I'm not yet convinced that an immediately-deleted offer should be considered a successful operation.

> We don't want to fail operations if they are "well formed", that way people can submit batches of trades (multiple ops) in one transaction and don't have to worry about some of those trades being only partial and/or fall victim of rounding.

Essentially, you're making a trade-off prioritizing dev usability (adding multiple ops per tx) over weird edge cases (any of those ops not hitting the orderbook). But you can make an argument for the reverse: that we should prioritize dev usability (understanding that success === the op did what I asked) over weird edge cases (the dev has to analyze the tx response + read documentation to figure out what happened).

I would argue that the second scenario is more insidious. We already see that wallets are not checking this offer response metadata. stellar.expert doesn't mention it at all:

https://stellar.expert/explorer/public/tx/97928226466205696#97928226466205697

To your point...

> We don't want people to implement code that is 90% correct in SDKs, we want to encourage people to deal with the "edge cases" by surfacing them right away

It seems like right now, with the system the way it is, people's code is only 90% correct (because they're not checking the offer status). And there's not a more direct way of surfacing this edge case than by failing the transaction.

**Comment by morleyzhi** [comment]
If we keep things the way they are, my plan is to add some properties to the response of `StellarSdk.Server.submitTransaction` that makes the offer responses a lot clearer, something like:

```javascript
    return HorizonAxiosClient.post(
      URI(this.serverURL)
        .segment('transactions')
        .toString(),
      `tx=${tx}`,
      { timeout: SUBMIT_TRANSACTION_TIMEOUT }
    )
      .then((response) => Object.assign({}, response.data, {
        manageOfferResults: [
          { operationIndex: 0, didCrossBook: false, offersClaimed: [], wasImmediatelyDeleted: true }
        ]
      }))
```

...But I think this illustrates how not-straightforward this response is for the developer. They have to read and understand what each state means (which to be honest, I don't even fully understand just yet), then cross-reference the operation that failed, then add special error-handling code to surface that one of the operations "failed" while the others succeeded. That's a lot to ask for an edge case that has pennies at stake.

**Comment by MonsieurNicolas** [comment]
We should continue this discussion in parallel.

I am interested in seeing if there is a way to make this simpler.

The constraints that we have are
1. it's very hard to deprecate the existing contract.
2. if we change when to fail, we have to have a very clear definition of the situation as to avoid having smart contracts fail when they should not.
 
We cannot dismiss the effect of rounding as "just pennies": while we keep the rounding error bounded within 1% of the target price, depending on what is being traded, it can still be significant.

Failing the operation for the "no offer created" case doesn't solve the problem of clients having to check that the offer that they placed has the right amount. From what you are saying, code doesn't check for this right now. Any reason for this?

The type of things that are not too hard to do are to add a new success type with different data. But we would need to know what makes sense first.

**Comment by jonjove** [comment]
@vcarl and @morleyzhi I agree with you that implementing logic to parse the `ManageOfferResult` is not the responsibility of application developers — it should be implemented in Horizon and/or SDKs. The reality is that the behavior of `ManageOffer` is really complicated, so it should be expected that the result is also complicated. But the fact that this is currently pushed down to the application level does not mean that the protocol needs to be changed; rather, it means that Horizon and/or SDKs need improved support for handling `ManageOffer`. 

Under typical circumstances, I think a user of `ManageOffer` should be checking the total amount sold (the sum of `amountBought` for every `ClaimOfferAtom` in `offersClaimed`), the total amount bought (the sum of `amountSold` for every `ClaimOfferAtom` in `offersClaimed`), and the current state of the offer as specified by `ManageOfferSuccesResult.offer` (which will either reveal that the offer was deleted or will provide access to the current state of the offer if it was created or updated). This is the central information that describes what happened to the offer you just submitted: how much executed and how much remains. I’m not sure I understand why a typical user would need more information than this, since any other information would pertain to offers on the ledger that are owned by other accounts. If you care about this, then you should probably be streaming all updates to the offer book whether you submitted them or not (as mentioned by @MonsieurNicolas).

To be extremely explicit: even if we changed the case where an offer does not execute and is immediately deleted to be an error (which I absolutely do not think we should do), you should still be checking the above if you care to know about the state of the offer you submitted.

**Comment by vcarl** [comment]
> code doesn't check [that the offer that they placed has the right amount] right now. Any reason for this?

I had no idea that was the expectation. Additionally, if I've just placed an offer and am checking the success response after it's crossed the orderbook, it's not clear to me what action I would be expected to take. On StellarX, we had a stream open to an account's orders to track when they were partially or totally filled, we didn't use the result of a trade submission to determine that whatsoever. If that's the expectation, that information should be surfaced in the JSON, not left in XDR. However I think this is a separate conversation, we can take it to Keybase or a separate issue if need be.

> it should be implemented in Horizon and/or SDKs

This sounds reasonable. It seems makes sense to me that the protocol may accept these transactions as well-formed, for Horizon to expose information to allow the SDKs to efficiently inspect them. My only concern is how it would be raised.

If Horizon errors, then I don't think that your original intent for allowing people to batch-submit transactions is met. If the SDKs error, then that leaves the possibility for SDKs to behave inconsistently, which doesn't seem right. If this is treated as a success but information is given to allow clients to handle this case easily, then each client implementation will have to have boilerplate around this.

Personally, I think this should error from Horizon to reduce burden on clients for typical usage. I'm curious to hear your thoughts @jonjove @MonsieurNicolas.

**Comment by tomquisel** [comment]
Closing this in favor of https://github.com/stellar/go/issues/1094, which I think deals with the lion's share of this issue in a general way at the Horizon level.
